### PR TITLE
[GEP-15] Rotate SSH Keys

### PR DESCRIPTION
**Pull request by xrstf**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
This PR implements the logic described in https://github.com/gardener/gardener/blob/master/docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md#rotation-proposal

**Release note**:

```feature operator
Shoot SSH Keys are regularly rotated, with both the current and previous key being deployed onto each shoot node.
```

### CODE REVIEW

File: pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser/component.go [hunk 1]
```diff
@@ -46,8 +47,8 @@ func init() {
 }
 
 const (
-	pathScript       = "/var/lib/gardener-user/run.sh"
-	pathPublicSSHKey = "/var/lib/gardener-user-ssh.key"
```
Comment (rfranzke) [inline pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser/component.go:50]: Should we pro-actively delete this file on existing nodes? Otherwise, it will stay there forever until the node is replaced. Probably it doesn't harm and it could be acceptable to prevent additional code, though, WDYT?

File: pkg/operation/botanist/secrets.go [hunk 2]
```diff
@@ -244,6 +251,55 @@ func (b *Botanist) rotateKubeconfigSecrets(ctx context.Context, gardenerResource
 	return b.K8sGardenClient.Client().Patch(ctx, b.Shoot.Info, client.MergeFrom(oldObj))
 }
 
+func (b *Botanist) rotateSSHKeypairSecrets(ctx context.Context, gardenerResourceDataList *gardencorev1alpha1helper.GardenerResourceDataList) error {
+	var err error
+
+	// fetch the current SSH keypair
+	currentSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.SecretNameSSHKeyPair, Namespace: b.Shoot.SeedNamespace}}
+
+	if err = b.K8sSeedClient.Client().Get(ctx, client.ObjectKeyFromObject(currentSecret), currentSecret); client.IgnoreNotFound(err) != nil {
```
Comment (rfranzke) [inline pkg/operation/botanist/secrets.go]: If the secret is not found, can the following code lead to panics?

File: pkg/operation/botanist/secrets.go [hunk 3]
```diff
@@ -244,6 +251,55 @@ func (b *Botanist) rotateKubeconfigSecrets(ctx context.Context, gardenerResource
...
+
+		if oldSecret.CreationTimestamp.IsZero() {
+			if err = b.K8sSeedClient.Client().Create(ctx, oldSecret); err != nil {
+				return fmt.Errorf("failed to create old SSH keypair secret: %v", err)
+			}
+		} else {
+			if err = b.K8sSeedClient.Client().Patch(ctx, oldSecret, patch); err != nil {
+				return fmt.Errorf("failed to patch old SSH keypair secret: %v", err)
+			}
+		}
```
Comment (rfranzke) [inline pkg/operation/botanist/secrets.go]: What about using some helper function from https://github.com/gardener/gardener/blob/master/pkg/controllerutils/patch.go?

File: pkg/operation/botanist/wanted_secrets.go [hunk 4]
```diff
@@ -481,6 +481,13 @@ func (b *Botanist) generateWantedSecretConfigs(basicAuthAPIServer *secrets.Basic
 			UsedForSSH: true,
 		},
 
+		// Secret definition for ssh-keypair.old
+		&secrets.RSASecretConfig{
+			Name:       v1beta1constants.SecretNameOldSSHKeyPair,
+			Bits:       4096,
+			UsedForSSH: true,
+		},
```
Comment (rfranzke) [inline pkg/operation/botanist/wanted_secrets.go:483]: For my understanding, why is this here?

**Review by rfranzke — COMMENTED** [review]
The GEP talks about the fact that this SSH key pair rotation happens once in the maintenance time window. Do you want to tackle this as part of this PR? I think what would be needed is to extend https://github.com/gardener/gardener/blob/db7118fe6a9072e462b32fe9aea92bdc11c88bf2/pkg/controllermanager/controller/shoot/shoot_maintenance_control.go#L157-L198 with setting the new `operation` annotation that you just introduced. However, this might conflict with https://github.com/gardener/gardener/blob/db7118fe6a9072e462b32fe9aea92bdc11c88bf2/pkg/controllermanager/controller/shoot/shoot_maintenance_control.go#L173-L175, so maybe it should only be done if the shoot does not need a retry? cc @ialidzhikov

File: pkg/operation/botanist/wanted_secrets.go (hunk 4 above)
Comment (xrstf) [reply pkg/operation/botanist/wanted_secrets.go:483]: Without this, the SecretManager never considered `ssh-keypair.old` to be a secret that it should sync back up to the garden. And I was under the impression that both SSH keys should be on the garden. It doesn't make sense to have a RSAConfig here, because I do not want a new key to be generated from scratch, but it was the only way I found to make both keys appear on the garden.

File: pkg/operation/botanist/secrets.go (hunk 2 above)
Comment (xrstf) [reply pkg/operation/botanist/secrets.go]: Nope, `oldPrivateKey := currentSecret.Data[secrets.DataKeyRSAPrivateKey]` won't panic and the following code doesn't either. Just tested by deleting all SSH keypair secrets in garden/seed and the gardenlet happily reconciled the state.

**Comment by xrstf** [comment]
> Do you want to tackle this as part of this PR?

Yes, that was my intention. We'll see how far I get :-D

File: pkg/operation/botanist/wanted_secrets.go (hunk 4 above)
Comment (rfranzke) [reply pkg/operation/botanist/wanted_secrets.go:483]: OK, thanks. Fine for me for now. cc @plkokanov as initial contributor of this secrets manager: I think we should simplify this soon because it's somewhat confusing and complex

File: pkg/operation/botanist/secrets.go (hunk 2 above)
Comment (rfranzke) [reply pkg/operation/botanist/secrets.go]: 👍🏻 thanks!

File: pkg/operation/botanist/wanted_secrets.go (hunk 4 above)
Comment (xrstf) [reply pkg/operation/botanist/wanted_secrets.go:483]: Yeah if we keep this, there needs to be a comment on this explaining why it's there.

File: pkg/apis/core/v1beta1/constants/types_constants.go [hunk 5]
```diff
@@ -36,6 +36,10 @@ const (
 	// SecretNameSSHKeyPair is a constant for the name of a Kubernetes secret object that contains the SSH key pair
 	// (public and private key) that can be used to SSH into the shoot nodes.
 	SecretNameSSHKeyPair = "ssh-keypair"
+	// SecretNameOldSSHKeyPair is asdasdsada constant for the name of a Kubernetes secret object that contains the previous
```
Comment (petersutter) [inline pkg/apis/core/v1beta1/constants/types_constants.go]:
```suggestion
	// SecretNameOldSSHKeyPair is a constant for the name of a Kubernetes secret object that contains the previous
```

File: pkg/operation/botanist/wanted_secrets.go (hunk 4 above)
Comment (xrstf) [reply pkg/operation/botanist/wanted_secrets.go:483]: To further clarify: With the current code, each new shoot gets 2 SSH keypairs immediately, even before the first rotation. It ain't pretty, but that's how far I got until now. ^^ I was hoping we can consider this a ... happy little accident. 🎨

**Comment by xrstf** [comment]
> so maybe it should only be done if the shoot does not need a retry?

I guess in a failed shoot, rotating keys doesn't make sense, so I would agree, only rotate if no retry is required. Of all possible operations, rotating the SSH keypairs is probably the least important.

**Comment by xrstf** [comment]
So this seems to work just fine for the most parts. I was able to SSH onto a shoot node after the keys had been rotated, so the path through the OperatingSystemConfig and stuff seems to work as well.

However, _existing_ nodes did not update their SSH keys. I had to create a completely new node. Not sure what's causing that problem.

File: pkg/operation/botanist/secrets.go [hunk 6]
```diff
@@ -244,6 +251,50 @@ func (b *Botanist) rotateKubeconfigSecrets(ctx context.Context, gardenerResource
...
+	currentSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.SecretNameSSHKeyPair, Namespace: b.Shoot.SeedNamespace}}
+
+	if err = b.K8sSeedClient.Client().Get(ctx, client.ObjectKeyFromObject(currentSecret), currentSecret); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("failed to get SSH keypair secret: %v", err)
+	}
+
+	oldPrivateKey := currentSecret.Data[secrets.DataKeyRSAPrivateKey]
+	oldPublicKey := currentSecret.Data[secrets.DataKeySSHAuthorizedKeys]
+
+	// if the is a current key, transfer it to the .old Secret
```
Comment (petersutter) [inline pkg/operation/botanist/secrets.go]:
```suggestion
	// if there is a current key, transfer it to the .old Secret
```

File: pkg/apis/core/v1beta1/constants/types_constants.go (hunk 5 above)
Comment (xrstf) [reply pkg/apis/core/v1beta1/constants/types_constants.go]: WTF happened there?!

File: pkg/operation/botanist/wanted_secrets.go (hunk 4 above)
Comment (plkokanov) [reply pkg/operation/botanist/wanted_secrets.go:483]: Sorry for commenting so late, the flow for the secrets generation currently is the following:
In the GenerateSecrets step it will get all secret configs defined in `generateWantedSecretConfigs` generate the secret metadata for each and save that in the `ShootState`. 
Then in the DeploySecrets step it will again get all required secrets from `generateWantedSecretConfigs` and for each secret configuration it will get the corresponding metadata from the `ShootState`, combine it with the static data that comes in the secret config and deploy a `Secret` object in the shoot's control plane.

I can spend some time to see if there is a way to avoid having both the old and new ssh configs added in `generateWantedSecretConfigs`. 

As for the whole secret management I have a couple of ideas on how to simplify it and add support for rotating credentials that would require keeping the old secret (e.g. etcd encryption which is currently done in separate  steps)

File: pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser/component.go (hunk 1 above)
Comment (xrstf) [reply pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser/component.go:50]: Yeah let's just delete the leftovers.

**Comment by xrstf** [comment]
> However, existing nodes did not update their SSH keys. I had to create a completely new node. Not sure what's causing that problem.

That should have been caused by a bug in the gardeneruser.sh, which would never update an `authorized_keys` file, only write it once when the node initially booted. Letting my Gardener installation now wait and see if it fixes itself.

**Review by petersutter — DISMISSED** [review]
/lgtm 👍 

the release note block header needs to be fixed. It has a header of `other feature` (which is both a category)
Format of the block header should be: `<category> <target_group>`
Maybe `feature operator` suits better?

File: pkg/operation/botanist/secrets.go [hunk 7]
```diff
@@ -244,6 +251,50 @@ func (b *Botanist) rotateKubeconfigSecrets(ctx context.Context, gardenerResource
...
+
+	oldPrivateKey := currentSecret.Data[secrets.DataKeyRSAPrivateKey]
+	oldPublicKey := currentSecret.Data[secrets.DataKeySSHAuthorizedKeys]
+
+	// if there is a current key, transfer it to the .old Secret
+	if len(oldPrivateKey) > 0 && len(oldPublicKey) > 0 {
+		oldSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.SecretNameOldSSHKeyPair, Namespace: b.Shoot.SeedNamespace}}
+		if err = b.K8sSeedClient.Client().Get(ctx, client.ObjectKeyFromObject(oldSecret), oldSecret); client.IgnoreNotFound(err) != nil {
+			return fmt.Errorf("failed to get old SSH keypair secret: %v", err)
+		}
```
Comment (rfranzke) [inline pkg/operation/botanist/secrets.go]: Why is this needed when you call `GetAndCreateOrMergePatch` next?

**Review by rfranzke — CHANGES_REQUESTED** [review]
Looks good, but I think https://github.com/gardener/gardener/blob/b72ac4af0c9d00454be8b75eba00d7eaf8df8b48/pkg/registry/core/shoot/strategy.go#L96-L100 should be adapted as well

File: pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser/templates/scripts/reconcile.tpl.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+id $USERNAME || useradd $USERNAME -mU
+
+# copy authorized_keys file
+mkdir -p $DIR_SSH
+cp -f "{{ .pathAuthorizedSSHKeys }}" $PATH_AUTHORIZED_KEYS
+chown $USERNAME:$USERNAME $PATH_AUTHORIZED_KEYS
+
+# remove unused legacy file
+if [ -f "{{ .pathPublicSSHKey }}" ]; then
+  rm -- "{{ .pathPublicSSHKey }}"
```
Comment (vpnachev) [inline pkg/operation/botanist/component/extensions/operatingsystemconfig/original/components/gardeneruser/templates/scripts/reconcile.tpl.sh]: I have never seen this syntax so far. Reading the man it is used when the filename starts with `-`, but here this is not the case. 
```suggestion
  rm -f "{{ .pathPublicSSHKey }}"
```

File: pkg/operation/botanist/secrets.go [hunk 9]
```diff
@@ -244,6 +251,50 @@ func (b *Botanist) rotateKubeconfigSecrets(ctx context.Context, gardenerResource
...
 
+func (b *Botanist) rotateSSHKeypairSecrets(ctx context.Context, gardenerResourceDataList *gardencorev1alpha1helper.GardenerResourceDataList) error {
+	var err error
+
+	// fetch the current SSH keypair
+	currentSecret := &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.SecretNameSSHKeyPair, Namespace: b.Shoot.SeedNamespace}}
+
+	if err = b.K8sSeedClient.Client().Get(ctx, client.ObjectKeyFromObject(currentSecret), currentSecret); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("failed to get SSH keypair secret: %v", err)
+	}
```
Comment (vpnachev) [inline pkg/operation/botanist/secrets.go]: I am wondering why you are not getting the secret from the botanist? I mean something like:
```suggestion
	currentSecret, ok := b.Secrets[v1beta1constants.SecretNameSSHKeyPair]

	if !ok {
		return fmt.Errorf("failed to get SSH keypair secret: %v", err)
	}
```

**Review by vpnachev — COMMENTED** [review]
Could you explain how this rotation is working with the system default users?
Each OS has its own default user, e.g. `core`, `admin`, `ec2-user`, etc. These users get their ssh keys during VM creation. 
From what I see in this PR, it is taking care only to rotate the ssh key for the `gardener` user.
Additionally, on GCP the ssh keys are fetched dynamically from the metadata service, how is the key rotate there?

**Comment by xrstf** [comment]
@vpnachev 

> From what I see in this PR, it is taking care only to rotate the ssh key for the gardener user.
> Additionally, on GCP the ssh keys are fetched dynamically from the metadata service, how is the key rotate there?

I kept to the GEP-15, which detailed how the Secret was supposed to be rotated. I wasn't even aware that there are different sources for SSH keys with different cloud providers ^^ and I'm not sure how to handle those.

File: pkg/operation/botanist/secrets.go (hunk 9 above)
Comment (xrstf) [reply pkg/operation/botanist/secrets.go]: There was no special reason, I was simply not super familiarwith how the `Secrets` map is constructred and used during the botanisting phase. Your suggestion is of course the right approach.

**Comment by vpnachev** [comment]
> @vpnachev
> 
> > From what I see in this PR, it is taking care only to rotate the ssh key for the gardener user.
> > Additionally, on GCP the ssh keys are fetched dynamically from the metadata service, how is the key rotate there?
> 
> I kept to the GEP-15, which detailed how the Secret was supposed to be rotated. I wasn't even aware that there are different sources for SSH keys with different cloud providers ^^ and I'm not sure how to handle those.


I guess we have to update the GEP and include this use-case there as well. 
Then the rotation can be improved later with other PR(s), but for now we should be aware about this limitation.

File: docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md [hunk 10]
```diff
@@ -222,3 +222,8 @@ Currently, the `ssh` key pair for the shoot nodes are created once during shoot
         - The "execution" script includes also the original user data script, which it writes to `PATH_CLOUDCONFIG`, compares it against the previous cloud config and runs the script in case it has changed
         - Running the [original user data](https://github.com/gardener/gardener/tree/master/pkg/operation/botanist/component/extensions/operatingsystemconfig/original) script will also run the `gardeneruser` component, where the `authorized_keys` file will be updated
         - After the most recent cloud-config user data was applied, the "execution" script annotates the node with `checksum/cloud-config-data: <cloud-config-checksum>` to indicate the success
+
+### Limitations
+
+For some cloud providers, like GCP, SSH keypairs are managed at the provider side and not on the seed cluster as Kubernetes
```
Comment (vpnachev) [inline docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md]: This is not correct. The SSH keys are always stored in the seed clusters as secrets and also synced to the garden cluster. 

Simply, on GCP the VMs do not have any static users and there is an agent on the nodes that syncs the users with their ssh keys from the GCP IAM service.

File: pkg/operation/botanist/secrets.go [hunk 11]
```diff
@@ -243,6 +250,48 @@ func (b *Botanist) rotateKubeconfigSecrets(ctx context.Context, gardenerResource
 	return b.K8sGardenClient.Client().Patch(ctx, b.Shoot.Info, client.MergeFrom(oldObj))
 }
 
+func (b *Botanist) rotateSSHKeypairSecrets(ctx context.Context, gardenerResourceDataList *gardencorev1alpha1helper.GardenerResourceDataList) error {
+	var err error
+
+	currentSecret, ok := b.Secrets[v1beta1constants.SecretNameSSHKeyPair]
```
Comment (plkokanov) [inline pkg/operation/botanist/secrets.go]: This will not work, since `b.Secrets` is populated in the `DeploySecrets` function.

I think you should use the `gardenerResourceDataList` to get the current secret and handle the whole rotation. Otherwise some changes will not get saved in the `ShootState`. 

You can try something like this (notice that I delete both the old and new ssh keypair secrets from the controlplane, this is so that the old secret gets redeployed with data from the shootstate):
```go
currentSecret := gardenerResourceDataList.Get(v1beta1constants.SecretNameSSHKeyPair)
if currentSecret == nil {
	return fmt.Errorf("no Secret named %s loaded", v1beta1constants.SecretNameSSHKeyPair)
}

oldSecret := currentSecret.DeepCopy()
oldSecret.Name = v1beta1constants.SecretNameOldSSHKeyPair
gardenerResourceDataList.Upsert(oldSecret)

secrets := []string{
	v1beta1constants.SecretNameSSHKeyPair,
	v1beta1constants.SecretNameOldSSHKeyPair,
}

for _, secretName := range secrets {
	if err := b.K8sSeedClient.Client().Delete(ctx, &corev1.Secret{ObjectMeta: metav1.ObjectMeta{Name: secretName, Namespace: b.Shoot.SeedNamespace}}); client.IgnoreNotFound(err) != nil {
		return err
	}
}
gardenerResourceDataList.Delete(v1beta1constants.SecretNameSSHKeyPair)

// remove operation annotation
```

**Review by plkokanov — CHANGES_REQUESTED** [review]

File: docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md (hunk 10 above)
Comment (xrstf) [reply docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md]: Perfect, let me just quote you directly. :)

**Comment by rfranzke** [comment]
> Looks good, but I think
> 
> https://github.com/gardener/gardener/blob/b72ac4af0c9d00454be8b75eba00d7eaf8df8b48/pkg/registry/core/shoot/strategy.go#L96-L100
> 
> should be adapted as well

@xrstf This is not yet addressed

File: docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md (hunk 10 above)
Comment (petersutter) [reply docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md]: I think this is missing the part where it mentions that the SSH keys of the default OS users are currently not rotated.

> Each OS has its own default user, e.g. core, admin, ec2-user, etc. These users get their ssh keys during VM creation.

**Comment by xrstf** [comment]
@rfranzke addressed now. Sorry, didn't see that :flushed:

**Review by rfranzke — DISMISSED** [review]
/lgtm

**Comment by timebertt** [comment]
Should we get this in or does anyone else need to take another round of review?

**Comment by petersutter** [comment]
> Should we get this in or does anyone else need to take another round of review?

https://github.com/gardener/gardener/pull/4224#discussion_r661369982 this is still missing / was not yet dismissed

File: docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md [hunk 12]
```diff
@@ -222,3 +222,7 @@ Currently, the `ssh` key pair for the shoot nodes are created once during shoot
         - The "execution" script includes also the original user data script, which it writes to `PATH_CLOUDCONFIG`, compares it against the previous cloud config and runs the script in case it has changed
         - Running the [original user data](https://github.com/gardener/gardener/tree/master/pkg/operation/botanist/component/extensions/operatingsystemconfig/original) script will also run the `gardeneruser` component, where the `authorized_keys` file will be updated
         - After the most recent cloud-config user data was applied, the "execution" script annotates the node with `checksum/cloud-config-data: <cloud-config-checksum>` to indicate the success
+
+### Limitations
+
```
Comment (petersutter) [inline docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md:227]:
```suggestion
### Limitations

Each operating system has its own default user (e.g. `core`, `admin`, `ec2-user` etc). These users get their ssh keys during VM creation (however there is a different handling on Google Cloud Platform as stated below). These keys currently do not get rotated respectively are not removed from the `authorized_keys` file. This means, that the initial `ssh` key will still be valid for the default operating system user.

```

@vpnachev is this correct?

Comment (vpnachev) [reply docs/proposals/15-manage-bastions-and-ssh-key-pair-rotation.md:227]: More or less yes. 

GCP is really different in how the OS users are manged. There is even no default user, except the `root`. But I like the they it is written now and it should not dig into so much details of how GCP is doing it.

**Review by petersutter — APPROVED** [review]
/lgtm, thanks!

**Review by rfranzke — APPROVED** [review]

**Comment by petersutter** [comment]
> the release note block header needs to be fixed. It has a header of `other feature` (which is both a category)
> Format of the block header should be: `<category> <target_group>`
> Maybe `feature operator` suits better?

As it was still wrong I have fixed it to `feature operator`
